### PR TITLE
ServiceBasedGcsClient support detect gcs server availability and retry

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
ServiceBasedGcsClient support detect gcs server availability and retry.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 1]
```diff
@@ -16,16 +16,30 @@ Status ServiceBasedJobInfoAccessor::AsyncAdd(
...
-      [job_id, data_ptr, callback](const Status &status, const rpc::AddJobReply &reply) {
-        if (callback) {
-          callback(status);
-        }
-        RAY_LOG(DEBUG) << "Finished adding job, status = " << status
-                       << ", job id = " << job_id
-                       << ", driver pid = " << data_ptr->driver_pid();
-      });
+
+  Executor* executor = new Executor(client_impl_);
```
Comment (zhijunfu) [inline src/ray/gcs/gcs_client/service_based_accessor.cc]: use smart pointers?

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 2]
```diff
@@ -16,16 +16,30 @@ Status ServiceBasedJobInfoAccessor::AsyncAdd(
...
+            RAY_LOG(DEBUG) << "Finished adding job, status = " << status
+                           << ", job id = " << job_id
+                           << ", driver pid = " << data_ptr->driver_pid();
+            delete executor;
+          } else {
+            RAY_LOG(WARNING) << "Failed to add job, retry again, status = " << status
+                          << ", job id = " << job_id
+                          << ", driver pid = " << data_ptr->driver_pid();
+            executor->post_execute(status);
+          }
```
Comment (zhijunfu) [inline src/ray/gcs/gcs_client/service_based_accessor.cc]: Is this code related to this PR? or is it fixing another issue?

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 3]
```diff
@@ -52,6 +52,32 @@ void ServiceBasedGcsClient::Disconnect() {
   RAY_LOG(INFO) << "ServiceBasedGcsClient Disconnected.";
 }
 
+uint64_t ServiceBasedGcsClient::Reconnect(uint64_t reconnect_count) {
+  absl::MutexLock lock(&mutex_);
```
Comment (zhijunfu) [inline src/ray/gcs/gcs_client/service_based_gcs_client.cc]: How is this function supposed to be called? And should this be async instead?

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: This code is related to this PR. Executor will check rpc IOError and reconnect gcs server and retry rpc operation.

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: Executor will be delete if rpc operation finished. If rpc operation failed,  we will use executor retry.

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc (hunk 3 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: Gcs client will be called by multiple threads, but it only has one rpc client, so sync is better.

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc [hunk 4]
```diff
@@ -52,6 +52,32 @@ void ServiceBasedGcsClient::Disconnect() {
...
+  if (reconnect_count_ >= reconnect_count) {
+    int num_attempts = 0;
+    while (num_attempts < RayConfig::instance().gcs_service_rpc_connect_retries()) {
```
Comment (micafan) [inline src/ray/gcs/gcs_client/service_based_gcs_client.cc]: Should reconnect in event loop (asynchronously) rather than occupy the thread.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 5]
```diff
@@ -16,16 +16,30 @@ Status ServiceBasedJobInfoAccessor::AsyncAdd(
...
+            }
+            RAY_LOG(DEBUG) << "Finished adding job, status = " << status
+                           << ", job id = " << job_id
+                           << ", driver pid = " << data_ptr->driver_pid();
+            delete executor;
+          } else {
+            RAY_LOG(WARNING) << "Failed to add job, retry again, status = " << status
+                          << ", job id = " << job_id
+                          << ", driver pid = " << data_ptr->driver_pid();
+            executor->post_execute(status);
```
Comment (micafan) [inline src/ray/gcs/gcs_client/service_based_accessor.cc]: We should pending all request when reconnecting, and keep requests in order if necessary.
The retry is invalid when the connection is not ok. And it may costs cpu.

File: src/ray/gcs/gcs_client.h [hunk 6]
```diff
@@ -60,6 +60,9 @@ class GcsClient : public std::enable_shared_from_this<GcsClient> {
   /// Disconnect with GCS Service. Non-thread safe.
   virtual void Disconnect() = 0;
 
+  /// Reconnect to GCS Service. Non-thread safe.
+  void Reconnect() {};
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client.h]: fixed

File: src/ray/common/ray_config_def.h [hunk 7]
```diff
@@ -234,3 +234,5 @@ RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
 /// Note: this only takes effect when gcs service is enabled.
 RAY_CONFIG(int64_t, gcs_service_connect_retries, 50)
 RAY_CONFIG(int64_t, gcs_service_connect_wait_milliseconds, 100)
+RAY_CONFIG(int64_t, gcs_service_rpc_connect_retries, 500)
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/common/ray_config_def.h]: fixed

File: src/ray/common/ray_config_def.h [hunk 8]
```diff
@@ -234,3 +234,5 @@ RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
 /// Note: this only takes effect when gcs service is enabled.
 RAY_CONFIG(int64_t, gcs_service_connect_retries, 50)
 RAY_CONFIG(int64_t, gcs_service_connect_wait_milliseconds, 100)
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/common/ray_config_def.h]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.h [hunk 9]
```diff
@@ -2,11 +2,34 @@
 #define RAY_GCS_SERVICE_BASED_ACCESSOR_H
 
 #include "src/ray/gcs/accessor.h"
+#include "src/ray/gcs/gcs_client/service_based_gcs_client.h"
 #include "src/ray/gcs/subscription_executor.h"
 
 namespace ray {
 namespace gcs {
 
+class Executor {
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.h]: fixed

File: src/ray/gcs/gcs_client/service_based_gcs_client.cc (hunk 4 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_gcs_client.cc]: fixed

**Review by zhijunfu — CHANGES_REQUESTED** [review]
I suggest we create a RPC client every time for reconnection, instead of putting the reconnection logic in RPC client for now, this should be simpler and cleaner, and also the reconnection logic can be complex and might vary in diffferent cases, for example, how to handle pending requests that haven't got replies, whether we need to resubmit them or just treat as failed, how many times we can reconnect, and what's the timeout interval for each reconnection, .etc. 

For now how about put the specific auto-reconnection logic to GCS related code, and create a new RPC client each time we want to reconnect like what core worker does it today for actor reconstruction. Later when we have a generic solution for reconnection we can consider adding it to RPC layer.

Does this make sense to you?

File: src/ray/common/ray_config_def.h [hunk 10]
```diff
@@ -237,7 +237,7 @@ RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
 /// Allow up to 5 seconds for connecting to gcs service.
 /// Note: this only takes effect when gcs service is enabled.
 RAY_CONFIG(int64_t, gcs_service_connect_retries, 50)
-RAY_CONFIG(int64_t, gcs_service_connect_wait_milliseconds, 100)
+RAY_CONFIG(int64_t, internal_gcs_service_connect_wait_milliseconds, 100)
```
Comment (zhijunfu) [inline src/ray/common/ray_config_def.h]: document the purpose of this config？

File: src/ray/rpc/gcs_server/gcs_rpc_client.h [hunk 11]
```diff
@@ -15,118 +61,143 @@ class GcsRpcClient {
...
-        new GrpcClient<StatsGcsService>(address, port, client_call_manager));
-    error_info_grpc_client_ = std::unique_ptr<GrpcClient<ErrorInfoGcsService>>(
-        new GrpcClient<ErrorInfoGcsService>(address, port, client_call_manager));
-    worker_info_grpc_client_ = std::unique_ptr<GrpcClient<WorkerInfoGcsService>>(
-        new GrpcClient<WorkerInfoGcsService>(address, port, client_call_manager));
+               ClientCallManager &client_call_manager,
+               std::function<std::pair<std::string, int>()> get_server_address = nullptr)
+      : client_call_manager_(client_call_manager),
+        get_server_address_(std::move(get_server_address)) {
+    init(address, port, client_call_manager);
```
Comment (zhijunfu) [inline src/ray/rpc/gcs_server/gcs_rpc_client.h]:
```suggestion
    Init(address, port, client_call_manager);
```

**Review by zhijunfu — COMMENTED** [review]
LGTM, a few minor comments

File: src/ray/rpc/gcs_server/gcs_rpc_client.h (hunk 11 above)
Comment (ffbin) [reply src/ray/rpc/gcs_server/gcs_rpc_client.h]: fixed

File: src/ray/common/ray_config_def.h (hunk 10 above)
Comment (ffbin) [reply src/ray/common/ray_config_def.h]: fixed

File: src/ray/rpc/grpc_client.h [hunk 12]
```diff
@@ -82,15 +88,38 @@ class GrpcClient {
...
+    std::pair<std::string, int> address = get_server_address_();
+    if (address_ != address) {
+      address_ = address;
+      std::shared_ptr<grpc::Channel> channel =
+          grpc::CreateCustomChannel(address.first + ":" + std::to_string(address.second),
+                                    grpc::InsecureChannelCredentials(), argument_);
+      stub_ = GrpcService::NewStub(channel);
+    }
+    return stub_;
+  }
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/rpc/grpc_client.h]: fixed

File: src/ray/common/ray_config_def.h [hunk 13]
```diff
@@ -237,7 +237,8 @@ RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
 /// Allow up to 5 seconds for connecting to gcs service.
 /// Note: this only takes effect when gcs service is enabled.
 RAY_CONFIG(int64_t, gcs_service_connect_retries, 50)
-RAY_CONFIG(int64_t, gcs_service_connect_wait_milliseconds, 100)
+/// Waiting time for each gcs service connection.
```
Comment (zhijunfu) [inline src/ray/common/ray_config_def.h]:
```suggestion
/// Waiting time for each gcs service reconnection.
```

**Review by zhijunfu — APPROVED** [review]
thanks, LGTM!
